tls/s2n_async_pkey.c

```diff
@@ -279,7 +282,16 @@ S2N_RESULT s2n_async_pkey_sign_sync(struct s2n_connection *conn, s2n_signature_a
     RESULT_GUARD(s2n_pkey_size(pkey, &maximum_signature_length));
     RESULT_GUARD_POSIX(s2n_alloc(&signed_content, maximum_signature_length));
 
-    RESULT_GUARD_POSIX(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
+    RESULT_ENSURE_REF(conn->config);
+    if (conn->config->verify_signatures) {
+        DEFER_CLEANUP(struct s2n_hash_state digest_for_verify, s2n_hash_free);
+        RESULT_GUARD_POSIX(s2n_hash_new(&digest_for_verify));
+        RESULT_GUARD_POSIX(s2n_hash_copy(&digest_for_verify, digest));
```

Comment:
Does sign or verify mutate the hash? Is that why you're copying it?

Reply:
Both. Calculating the digest of a hash "finalizes" it, so it can't be reused. We need a copy for sign and a copy for digest.

That's the standard model for hashes/digests. You create one, update it until it contains all your data, and then call digest/finalize to actually produce the final digest bytes. But that means that if you want to hash the same data twice, you have to copy the hash before you finalize it.

We can't just compute the digest once and pass the raw bytes to both the signing and verifying methods, because that's a whole can of worms. Our pkey operations hide a lot of messy logic. For one, the newer FIPS standards require that the digest/sign or digest/verify happen entirely inside the FIPS boundary, meaning that we can't just pass in raw digest bytes because the libcrypto can't validate those raw bytes are actually the digest.